pennylane/transforms/decompositions/two_qubit_unitary.py

```diff
@@ -0,0 +1,346 @@
+    C2 = U[0:2, 2:4]
+    C3 = U[2:4, 0:2]
+    C4 = U[2:4, 2:4]
+
+    # From the definition of A \otimes B, C1 C4^\dag = a1^2 I, so we can extract a1
+    C14 = qml.math.dot(C1, qml.math.conj(qml.math.T(C4)))
+    a1 = qml.math.sqrt(C14[0, 0])
+
+    # Similarly, -C2 C3^\dag = a2^2 I, so we can extract a2
+    C23 = qml.math.dot(C2, qml.math.conj(qml.math.T(C3)))
+    a2 = qml.math.sqrt(-C23[0, 0])
+
+    # This gets us a1, a2 up to a sign. To resolve the sign, ensure that
+    # C1 C2^dag = a1 a2* I
+    C12 = qml.math.dot(C1, qml.math.conj(qml.math.T(C2)))
+
+    if not qml.math.isclose(a1 * np.conj(a2), C12[0, 0]):
+        a2 *= -1
+
+    # Construct A
+    A = qml.math.stack([[a1, a2], [-qml.math.conj(a2), qml.math.conj(a1)]])
+
+    # Next, extract B. Can do from any of the C, just need to be careful in
+    # case one of the elements of A is 0.
+    if not qml.math.isclose(A[0, 0], 0.0, atol=1e-8):
+        B = C1 / A[0, 0]
+    else:
+        B = C2 / A[0, 1]
+
+    return A, B
+
+
+def _extract_su2su2_prefactors(U, V):
+    """U, V are SU(4) matrices for which there exists A, B, C, D such that
+    (A \otimes B) V (C \otimes D) = U. The problem is to find A, B, C, D in SU(2)
+    in an analytic and fully differentiable manner.
+
+    This decomposition is possible when U and V are in the same double coset of
+    SU(4), meaning there exists G, H in SO(4) s.t. G (Edag V E) H = (Edag U
+    E). This is guaranteed here by how V was constructed using the
+    _select_rotation_angles method. Then, we can use the fact that E SO(4) Edag
+    gives us something in SU(2) x SU(2) to give A, B, C, D.
+    """
+
+    # A lot of the work here happens in the magic basis. Essentially, we
+    # don't look explicitly at some U = G V H, but rather at
+    #     E^\dagger U E = G E^\dagger V E H
+    # so that we can recover
+    #     U = (E G E^\dagger) V (E H E^\dagger) = (A \otimes B) V (C \otimes D).
+    # There is some math in the paper explaining how when we define U in this way,
+    # we can simultaneously diagonalize functions of U and V to ensure they are
+    # in the same coset and recover the decomposition.
+
+    u = qml.math.linalg.multi_dot([Edag, U, E])
+    v = qml.math.linalg.multi_dot([Edag, V, E])
+
+    uuT = qml.math.dot(u, qml.math.T(u))
+    vvT = qml.math.dot(v, qml.math.T(v))
+
+    # First, we find a matrix p (hopefully) in SO(4) s.t. p^T u u^T p is diagonal.
+    # Since uuT is complex and symmetric, both its real / imag parts share a set
+    # of real-valued eigenvectors.
+    ev_p, p = qml.math.linalg.eig(uuT.real)
+
+    # We also do this for v, i.e., find q (hopefully) in SO(4) s.t. q^T v v^T q is diagonal.
+    ev_q, q = qml.math.linalg.eig(vvT.real)
+
+    # If determinant of p is not 1, it is in O(4) but not SO(4), and has
+    # determinant -1. We can transform it to SO(4) by simply negating one
+    # of the columns.
+    if not qml.math.isclose(qml.math.linalg.det(p), 1.0):
+        p[:, -1] = -p[:, -1]
```

Comment:
Ah, in-place array element setting isn't supported in autograd and TF (but does work in Torch!).

This will have to be re-written to avoid the array assignment. One solution is to 'unstack' the `p` matrix using `qml.math.unstack`, negating the last column, and then restacking.

Maybe a more elegant way is to do a matrix multiplication, or even just scalar multiplication?

Reply:
Good catch, should just be able to right-multiply by a matrix `np.diag([1, 1, 1, -1])` (or something like this) :+1: